### ISSUE TITLE
optimize sync's performed across three controller on Update to the objects

### ISSUE THREAD

**Issue by murali-reddy** [cloudnativelabs/kube-router#386]
API object changes for various reasons (update to status, change in meta data). We need to restrict the sync() performed by the controller on each Update(). Most of the time change does not necessarily warrants sync. 

For `update` function check if the update has a change that warrants the sync
```go
		UpdateFunc: func(old, new interface{}) {
			key, err := cache.MetaNamespaceKeyFunc(new)
			if err == nil {
				newEndpointsObj := new.(*v1core.Endpoints)
				oldEndpointsObj := old.(*v1core.Endpoints)
				if endpointsObjChanged(oldEndpointsObj, newEndpointsObj) {
					// take action
				}
			}
		},
```
it varies from API object to object. For e.g for endpoints object all we care is if subset has changes, any other changes to the endpoints object irrelevant

```go
func endpointsObjChanged(oldEndpointsObj, newEndpointsObj *v1core.Endpoints) bool {

	if !apiequality.Semantic.DeepEqual(newEndpointsObj.EndpointSubset, oldEndpointsObj.EndpointSubset)  {
		return true
	}
	return false
}
```

**Comment by andrewsykim** [comment]
@murali-reddy would be great if we can first optimize services controller to do a sync per service/endpoints. Right now it syncs all endpoints/services for all update events.

**Comment by murali-reddy** [comment]
Agree, there has to be full sync and delta sync.

**Comment by murali-reddy** [comment]
Fixed as part of #400 #399 

there is still some work left:

- use work queue
- on failure to act upon a update, reque the object to the work queue

Which will even reduce the need for periodic sync.
